File: packages/snow_draw_engine/lib/draw/elements/types/line/line_task_encoder.dart
import '../../../models/element_state.dart';
import '../../../render/tasks/render_tasks.dart';
import '../../../services/text/text_metrics_service.dart';
import '../../core/typed_element_render_task_encoder.dart';
import 'line_data.dart';

/// Encodes lines into high-level render tasks.
final class LineTaskEncoder extends TypedElementRenderTaskEncoder<LineData> {
  /// Creates a line task encoder.
  const LineTaskEncoder();

  @override
  List<RenderTask> encodeTypedTasks({
    required ElementState element,
    required LineData data,
    String? localeTag,
    TextMetricsService? textMetricsService,
  }) => <RenderTask>[
    LineRenderTask(element: element, data: data, localeTag: localeTag),
  ];
}

File: packages/snow_draw_rust_engine/src/draw/elements/types/line/line_task_encoder.rs
#![allow(dead_code)]

use std::sync::Arc;

use super::line_data::LineData;
use crate::draw::elements::core::typed_element_render_task_encoder::{
    ElementState, RenderTaskList, TypedElementRenderTaskEncoder,
};
use crate::draw::types::edit_context::TextMetricsService;

/// Element-level render task for line payloads.
///
/// Mirrors Dart `LineRenderTask(element: ..., data: ..., localeTag: ...)`.
#[derive(Debug, Clone)]
pub struct LineRenderTask {
    pub element: ElementState,
    pub data: LineData,
    pub locale_tag: Option<String>,
}

impl LineRenderTask {
    /// Creates a new line render task snapshot.
    pub fn new(element: ElementState, data: LineData, locale_tag: Option<String>) -> Self {
        Self {
            element,
            data,
            locale_tag,
        }
    }
}

/// Encodes lines into high-level render tasks.
///
/// Translation of Dart `LineTaskEncoder extends TypedElementRenderTaskEncoder<LineData>`.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct LineTaskEncoder;

impl LineTaskEncoder {
    /// Creates a line task encoder.
    pub const fn new() -> Self {
        Self
    }
}

impl TypedElementRenderTaskEncoder<LineData> for LineTaskEncoder {
    fn encode_typed_tasks(
        &self,
        element: &ElementState,
        data: &LineData,
        locale_tag: Option<&str>,
        _text_metrics_service: Option<Arc<dyn TextMetricsService>>,
    ) -> RenderTaskList {
        vec![Box::new(LineRenderTask::new(
            element.clone(),
            data.clone(),
            locale_tag.map(str::to_owned),
        ))]
    }
}
